File: src/main/java/io/github/arieslab/business/daos/utils/DAOGeneric.java
package io.github.arieslab.business.daos.utils;

import org.hibernate.Criteria;
import org.hibernate.Query;
import org.hibernate.SQLQuery;
import org.hibernate.SessionFactory;
import org.hibernate.criterion.Criterion;
import org.hibernate.criterion.Order;
import org.hibernate.criterion.Projections;
import org.springframework.beans.factory.annotation.Autowired;

import java.lang.reflect.ParameterizedType;
import java.util.List;
import java.util.Map;

@SuppressWarnings("unchecked")
public class DAOGeneric<T> {

    @Autowired
    private SessionFactory sessionFactory;
    private Class<T> persistentClass;

    public DAOGeneric() {
        this.persistentClass = (Class<T>) ((ParameterizedType) getClass().getGenericSuperclass()).getActualTypeArguments()[0];
    }

    /**
     * Sets the Hibernate SessionFactory.
     *
     * @param sf the SessionFactory
     */
    public void setSessionFactory(SessionFactory sf) {
        this.sessionFactory = sf;
    }

    /**
     * Returns the current Hibernate session.
     *
     * @return the current Session
     */
    public org.hibernate.Session session() {
        return sessionFactory.getCurrentSession();
    }

    /**
     * Returns the persistent class type.
     *
     * @return the entity class
     */
    public Class<T> clazz() {
        return this.persistentClass;
    }

    /**
     * Deletes an entity from the database.
     *
     * @param entity the entity to delete
     */
    public void delete(T entity) {
        session().delete(entity);
    }

    /**
     * Deletes an entity by its ID using a HQL query.
     *
     * @param id the entity ID
     */
    public void delete(Long id) {
        String hql = "delete " + clazz().getSimpleName() + " where id = :id";
        Query q = session().createQuery(hql).setParameter("id", id);
        q.executeUpdate();
    }

    /**
     * Finds an entity by its ID.
     *
     * @param id the entity ID
     * @return the found entity, or null
     */
    public T findById(Long id) {
        return (T) session().get(clazz(), id);
    }

    /**
     * Lists all entities of this type.
     *
     * @return list of all entities
     */
    public List<T> listAll() {
        return findByHQL("FROM " + clazz().getSimpleName());
    }

    /**
     * Returns the number of entities in the database.
     *
     * @return the entity count
     */
    public int size() {
        Criteria crit = session().createCriteria(clazz());
        crit.setProjection(Projections.rowCount());
        Object result = crit.uniqueResult();
        Long size = (Long) result;
        return size.intValue();
    }

    /**
     * Saves or updates an entity.
     *
     * @param entity the entity to save
     * @return the saved entity
     */
    public T save(T entity) {
        session().saveOrUpdate(entity);
        return entity;
    }

    /**
     * Finds entities matching the given Hibernate criteria.
     *
     * @param criterion the criteria to match
     * @return list of matching entities
     */
    public List<T> findByCriteriaReturnList(Criterion... criterion) {
        return findByCriteria(null, criterion);
    }

    /**
     * Finds a single entity matching the given criteria.
     *
     * @param criterion the criteria to match
     * @return the matching entity, or null
     */
    public T findByCriteriaReturnUniqueResult(Criterion... criterion) {
        return findByCriteriaReturnUniqueResult(null, criterion);
    }

    /**
     * Finds entities matching criteria with optional ordering.
     *
     * @param order optional ordering
     * @param criterion the criteria to match
     * @return list of matching entities
     */
    public List<T> findByCriteria(Order order, Criterion... criterion) {
        Criteria crit = session().createCriteria(clazz());
        crit.setResultTransformer(Criteria.DISTINCT_ROOT_ENTITY);
        for (Criterion c : criterion) {
            crit.add(c);
        }
        if (order != null) {
            crit.addOrder(order);
        }
        return crit.list();
    }

    /**
     * Finds entities matching criteria with ordering, limited to maxResult.
     *
     * @param orders list of ordering specifications
     * @param criterion list of criteria
     * @param maxResult maximum number of results
     * @return list of matching entities
     */
    public List<T> findByCriteria(List<Order> orders, List<Criterion> criterion, int maxResult) {
        Criteria crit = session().createCriteria(clazz());

        crit.setMaxResults(maxResult);

        crit.setResultTransformer(Criteria.DISTINCT_ROOT_ENTITY);
        for (Criterion c : criterion) {
            crit.add(c);
        }
        for (Order o : orders) {
            crit.addOrder(o);
        }
        return crit.list();
    }

    /**
     * Finds entities matching criteria with ordering and aliases.
     *
     * @param orders list of ordering specifications
     * @param criterion list of criteria
     * @param alias map of alias names
     * @return list of matching entities
     */
    public List<T> findByCriteria2(List<Order> orders, List<Criterion> criterion, Map<String, String> alias) {
        Criteria crit = session().createCriteria(clazz());

        if (alias != null) {
            for (Map.Entry<String, String> entry : alias.entrySet()) {
                crit.createAlias(entry.getKey(), entry.getValue());
            }
        }

        crit.setResultTransformer(Criteria.DISTINCT_ROOT_ENTITY);
        for (Criterion c : criterion) {
            crit.add(c);
        }
        for (Order o : orders) {
            crit.addOrder(o);
        }
        return crit.list();
    }

    /**
     * Finds a single entity matching criteria with optional ordering.
     *
     * @param order optional ordering
     * @param criterion the criteria to match
     * @return the matching entity, or null
     */
    public T findByCriteriaReturnUniqueResult(Order order, Criterion... criterion) {
        Criteria crit = session().createCriteria(clazz());
        crit.setResultTransformer(Criteria.DISTINCT_ROOT_ENTITY);
        for (Criterion c : criterion) {
            crit.add(c);
        }
        if (order != null) {
            crit.addOrder(order);
        }
        return (T) crit.uniqueResult();
    }

    /**
     * Finds a single entity with pagination and ordering.
     *
     * @param order optional ordering
     * @param offSet the offset
     * @param size the page size
     * @param criterion the criteria to match
     * @return the matching entity, or null
     */
    public T findByCriteriaReturnUniqueResult(Order order, int offSet, int size, Criterion... criterion) {
        Criteria crit = session().createCriteria(clazz());
        crit.setResultTransformer(Criteria.DISTINCT_ROOT_ENTITY);
        crit.setFirstResult(offSet);
        crit.setMaxResults(size);
        for (Criterion c : criterion) {
            crit.add(c);
        }
        if (order != null) {
            crit.addOrder(order);
        }
        return (T) crit.uniqueResult();
    }

    /**
     * Lists all entities with optional ordering.
     *
     * @param order optional ordering
     * @return list of entities
     */
    public List<T> findByCriteria(Order order) {
        Criteria crit = session().createCriteria(clazz());
        crit.setResultTransformer(Criteria.DISTINCT_ROOT_ENTITY);
        if (order != null) {
            crit.addOrder(order);
        }
        return crit.list();
    }

    /**
     * Lists entities with pagination and optional ordering.
     *
     * @param order optional ordering
     * @param offSet the offset
     * @param size the page size
     * @return list of entities
     */
    public List<T> findByCriteria(Order order, int offSet, int size) {
        Criteria crit = session().createCriteria(clazz());
        crit.setResultTransformer(Criteria.DISTINCT_ROOT_ENTITY);
        crit.setFirstResult(offSet);
        crit.setMaxResults(size);
        if (order != null) {
            crit.addOrder(order);
        }
        return crit.list();
    }

    /**
     * Executes a paginated HQL query with ordering.
     *
     * @param order the ordering
     * @param offSet the offset
     * @param size the page size
     * @return list of results
     */
    public List<T> findByHQL(Order order, int offSet, int size) {
        Query query = session().createQuery("FROM " + clazz().getSimpleName() + " order by " + order.getPropertyName());
        query.setFirstResult(offSet);
        query.setMaxResults(size);
        return query.list();
    }

    /**
     * Executes an HQL query.
     *
     * @param hql the HQL query string
     * @return list of results
     */
    public List<T> findByHQL(String hql) {
        Query query = session().createQuery(hql);
        return query.list();
    }

    /**
     * Executes an HQL query returning a single result.
     *
     * @param hql the HQL query string
     * @return the single result, or null
     */
    public T findByHQLUniqueResult(String hql) {
        Query query = session().createQuery(hql);
        return (T) query.uniqueResult();
    }

    /**
     * Executes a native SQL update query.
     *
     * @param query the SQL query string
     */
    public void executeSQL(String query) {
        session().createSQLQuery(query).executeUpdate();
    }

    /**
     * Executes a native SQL query and returns results as a list of maps.
     *
     * @param sql the SQL query string
     * @return list of result rows as maps
     */
    public List executeSQL_List(String sql) {
        SQLQuery q = session().createSQLQuery(sql);
        q.setResultTransformer(Criteria.ALIAS_TO_ENTITY_MAP);
        return q.list();
    }
}

File: src/main/java/io/github/arieslab/business/daos/utils/DAOGeneric.kt
package io.github.arieslab.business.daos.utils

import jakarta.persistence.NoResultException
import jakarta.persistence.TypedQuery
import jakarta.persistence.criteria.CriteriaBuilder
import jakarta.persistence.criteria.CriteriaQuery
import jakarta.persistence.criteria.Order
import jakarta.persistence.criteria.Predicate
import org.hibernate.Session
import org.hibernate.SessionFactory
import org.springframework.beans.factory.annotation.Autowired

open class DAOGeneric<T> {

    @Autowired
    private lateinit var sessionFactory: SessionFactory

    @Suppress("UNCHECKED_CAST")
    private val persistentClass: Class<T> by lazy {
        (javaClass.genericSuperclass as? java.lang.reflect.ParameterizedType)
            ?.actualTypeArguments?.get(0) as? Class<T>
            ?: throw RuntimeException("Could not determine persistent class")
    }

    fun setSessionFactory(sf: SessionFactory) {
        sessionFactory = sf
    }

    fun session(): Session = sessionFactory.currentSession

    fun clazz() = persistentClass

    fun delete(entity: T) {
        session().remove(entity)
    }

    fun delete(id: Long) {
        val entity = findById(id)
        if (entity != null) {
            session().remove(entity)
        }
    }

    fun findById(id: Long): T? = session().find(clazz(), id)

    fun listAll(): List<T> = findByHQL("FROM ${clazz().simpleName}")

    fun size(): Int {
        val cb = session().criteriaBuilder
        val query = cb.createQuery(Long::class.java)
        query.select(cb.count(query.from(clazz())))
        return session().createQuery(query).singleResult.toInt()
    }

    fun save(entity: T): T {
        session().persist(entity)
        return entity
    }

    fun findByCriteriaReturnList(vararg predicates: Predicate): List<T> {
        return findByCriteria(null as Order?, *predicates)
    }

    fun findByCriteriaReturnUniqueResult(vararg predicates: Predicate): T? {
        return findByCriteriaReturnUniqueResult(null as Order?, *predicates)
    }

    fun findByCriteria(order: Order?, vararg predicates: Predicate): List<T> {
        val cb = session().criteriaBuilder
        val query: CriteriaQuery<T> = cb.createQuery(clazz())
        val root = query.from(clazz())
        query.select(root).distinct(true)
        if (predicates.isNotEmpty()) {
            query.where(*predicates)
        }
        if (order != null) {
            query.orderBy(order)
        }
        return session().createQuery(query).resultList
    }

    fun findByCriteria(orders: List<Order>?, predicates: List<Predicate>?, maxResult: Int): List<T> {
        val cb = session().criteriaBuilder
        val query: CriteriaQuery<T> = cb.createQuery(clazz())
        val root = query.from(clazz())
        query.select(root).distinct(true)
        if (!predicates.isNullOrEmpty()) {
            query.where(*predicates.toTypedArray())
        }
        if (!orders.isNullOrEmpty()) {
            query.orderBy(orders)
        }
        val q = session().createQuery(query)
        if (maxResult > 0) {
            q.maxResults = maxResult
        }
        return q.resultList
    }

    fun findByCriteriaWithAliases(orders: List<Order>?, predicates: List<Predicate>?, alias: Map<String, String>?) {
        val cb = session().criteriaBuilder
        val query: CriteriaQuery<T> = cb.createQuery(clazz())
        val root = query.from(clazz())
        query.select(root).distinct(true)
        if (alias != null) {
            for ((key, value) in alias) {
                @Suppress("UNCHECKED_CAST")
                (root.join<Any, Any>(key) as Any).let { }
                root.join<Any, Any>(key).alias(value)
            }
        }
        if (!predicates.isNullOrEmpty()) {
            query.where(*predicates.toTypedArray())
        }
        if (!orders.isNullOrEmpty()) {
            query.orderBy(orders)
        }
    }

    fun findByCriteriaReturnUniqueResult(order: Order?, vararg predicates: Predicate): T? {
        val cb = session().criteriaBuilder
        val query: CriteriaQuery<T> = cb.createQuery(clazz())
        val root = query.from(clazz())
        query.select(root).distinct(true)
        if (predicates.isNotEmpty()) {
            query.where(*predicates)
        }
        if (order != null) {
            query.orderBy(order)
        }
        return try {
            session().createQuery(query).singleResult
        } catch (e: NoResultException) {
            null
        }
    }

    fun findByCriteriaReturnUniqueResult(order: Order?, offSet: Int, size: Int, vararg predicates: Predicate): T? {
        val cb = session().criteriaBuilder
        val query: CriteriaQuery<T> = cb.createQuery(clazz())
        val root = query.from(clazz())
        query.select(root).distinct(true)
        if (predicates.isNotEmpty()) {
            query.where(*predicates)
        }
        if (order != null) {
            query.orderBy(order)
        }
        val q = session().createQuery(query)
        q.firstResult = offSet
        q.maxResults = size
        return try {
            q.singleResult
        } catch (e: NoResultException) {
            null
        }
    }

    fun findByCriteria(order: Order?): List<T> {
        val cb = session().criteriaBuilder
        val query: CriteriaQuery<T> = cb.createQuery(clazz())
        val root = query.from(clazz())
        query.select(root).distinct(true)
        if (order != null) {
            query.orderBy(order)
        }
        return session().createQuery(query).resultList
    }

    fun findByCriteria(order: Order?, offSet: Int, size: Int): List<T> {
        val cb = session().criteriaBuilder
        val query: CriteriaQuery<T> = cb.createQuery(clazz())
        val root = query.from(clazz())
        query.select(root).distinct(true)
        if (order != null) {
            query.orderBy(order)
        }
        val q = session().createQuery(query)
        q.firstResult = offSet
        q.maxResults = size
        return q.resultList
    }

    fun findByHQL(orderBy: String, ascending: Boolean, offSet: Int, size: Int): List<T> {
        val dir = if (ascending) "asc" else "desc"
        val query = session().createQuery(
            "FROM ${clazz().simpleName} order by $orderBy $dir", clazz())
        query.firstResult = offSet
        query.maxResults = size
        return query.resultList
    }

    fun findByHQL(hql: String): List<T> {
        return session().createQuery(hql, clazz()).resultList
    }

    fun findByHQLUniqueResult(hql: String): T? {
        val query = session().createQuery(hql, clazz())
        return try {
            query.singleResult
        } catch (e: NoResultException) {
            null
        }
    }

    fun executeSQL(sql: String) {
        session().createNativeQuery(sql).executeUpdate()
    }

    @Suppress("unchecked_cast")
    fun executeSQL_List(sql: String): List<*> {
        return session().createNativeQuery(sql).resultList
    }
}
